judge.pl

```diff
@@ -303,6 +308,27 @@ sub validate_test {
     $sp_report->ok;
 }
 
+sub check_input_hash {
+    my ($pid, $rank, $filename) = @_;
+
+    eval { require Authen::Passphrase::SaltedDigest; 1; } or return;
+
+    my $data = $fu->read_lines($filename);
+    my $input = join "", @$data;
+
+    my $hash = $judge->get_input_hash($pid, $rank);
+
+    if (defined $hash) {
+        my $p = Authen::Passphrase::SaltedDigest->from_rfc2307($hash);
+        $p->match($input) or die "Invalid hash for test $rank";
+    }
+    else {
+        $hash = Authen::Passphrase::SaltedDigest->new(algorithm => "SHA-1",
+            passphrase => $input)->as_rfc2307;
+        $judge->save_input_hash($pid, $rank, $hash);
+    }
+}
+
```

Comment:
Sorry, I failed to save the most important comment during review.
It turns out that RFC2037 format in not suitable for us. It is old and frozen, does not extend to newer hashes,
and uses Base64 encoding which is bad for human-editable hashes in problem package.
Also, Authen::Passphrase is a quite heave dependence which should at least be added to installer,
but better yet not added altogether.
I suggest to manually create MFC/PHC-like format, but with hex string instead of Base64 (i.e. `$sha1$deadbeef...`).
Implement only SHA-1 and MD5, fail on bad format (extract format validation code to cats-problem and test it there).